File: .github/e2e/extract-results.py
#!/usr/bin/env python3
"""
Extract the test results JSON from a Claude Code stream-json output file.

Usage: extract-results.py <raw-output.txt>
Writes the results JSON object to stdout.
"""
import sys, json, re


def extract_json_with_overall(text):
    """Return the last JSON object containing an 'overall' key, or None."""
    # Prefer fenced ```json blocks (most reliable)
    for b in reversed(re.findall(r'```json\s*([\s\S]*?)```', text)):
        try:
            obj = json.loads(b.strip())
            if 'overall' in obj:
                return obj
        except Exception:
            pass
    # Fallback: depth-tracking scan for any {...} containing 'overall'
    depth = 0; start = -1; best = None
    for i, c in enumerate(text):
        if c == '{':
            if depth == 0:
                start = i
            depth += 1
        elif c == '}':
            depth -= 1
            if depth == 0 and start >= 0:
                try:
                    obj = json.loads(text[start:i + 1])
                    if 'overall' in obj:
                        best = obj
                except Exception:
                    pass
                start = -1
    return best


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else 'raw-output.txt'
    try:
        content = open(path, encoding='utf-8', errors='replace').read()
    except OSError:
        print('{"overall":"FAIL","error":"raw-output.txt not found"}')
        sys.exit(1)

    result_text = ''
    all_texts = []
    for line in content.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
        except Exception:
            continue
        if obj.get('type') == 'result' and obj.get('result'):
            result_text = obj['result']
            all_texts.append(result_text)
        elif obj.get('type') == 'assistant':
            for block in obj.get('message', {}).get('content', []):
                if block.get('type') == 'text' and block.get('text'):
                    all_texts.append(block['text'])

    # Try the final result text first, then all accumulated assistant text
    for text in ([result_text] if result_text else []) + ['\n'.join(all_texts)]:
        found = extract_json_with_overall(text)
        if found:
            print(json.dumps(found))
            return

    # Last resort: reconstruct from the most recent CHECKPOINT line
    combined = '\n'.join(all_texts)
    checkpoints = []
    for line in combined.splitlines():
        if line.startswith('CHECKPOINT: '):
            try:
                cp = json.loads(line[len('CHECKPOINT: '):])
                checkpoints.append(cp)
            except Exception:
                pass
    if checkpoints:
        last = checkpoints[-1]
        overall = 'FAIL' if any(t.get('status') == 'FAIL' for t in last) else 'PASS'
        print(json.dumps({'results': last, 'overall': overall,
                          'note': 'reconstructed from checkpoints'}))
        return

    print('{"overall":"FAIL","error":"no results JSON found in output"}')


if __name__ == '__main__':
    main()

File: .github/e2e/extract-results.mjs
#!/usr/bin/env node
import { readFileSync, existsSync, readdirSync } from 'node:fs';
import { join } from 'node:path';

const args = process.argv.slice(2);

if (args.length < 4) {
  process.stderr.write('Usage: extract-results.mjs <suite> <pm_os> <pm_provider> <raw-file1> [raw-file2 ...]\n');
  process.exit(1);
}

const [suite, pmOs, pmProvider, ...rawFiles] = args;

const runDir = rawFiles[0].split(/[\\/]/).slice(0, -1).join('/');

function processRawFile(filePath, provider) {
  let assistantText = '';
  let tokensIn = 0;
  let tokensOut = 0;
  let cacheCreate = 0;
  let cacheRead = 0;

  if (!existsSync(filePath)) {
    return { assistantText, tokensIn, tokensOut, cacheCreate, cacheRead };
  }

  const content = readFileSync(filePath, 'utf8');

  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    let obj;
    try { obj = JSON.parse(trimmed); } catch { continue; }

    if (provider === 'gemini') {
      // Gemini: accumulate from result event stats (input = non-cached input, cached = cache reads)
      if (obj.type === 'result' && obj.stats) {
        const s = obj.stats;
        tokensIn += (s.input ?? 0);
        tokensOut += (s.output_tokens ?? 0);
        cacheRead += (s.cached ?? 0);
        // cacheCreate stays 0: gemini does not report cache writes
      }
    } else {
      // Claude: sum usage across every assistant event (not just the final result turn)
      if (obj.type === 'assistant' && obj.message?.usage) {
        const u = obj.message.usage;
        tokensIn += (u.input_tokens ?? 0);
        tokensOut += (u.output_tokens ?? 0);
        cacheCreate += (u.cache_creation_input_tokens ?? 0);
        cacheRead += (u.cache_read_input_tokens ?? 0);
      }
    }

    if (obj.type === 'result' && obj.result) {
      assistantText += '\n' + obj.result;
    } else if (obj.type === 'assistant') {
      for (const block of obj.message?.content ?? []) {
        if (block?.type === 'text' && block.text) assistantText += '\n' + block.text;
      }
    } else if (obj.type === 'message' && obj.role === 'assistant' && typeof obj.content === 'string') {
      assistantText += obj.content;
    }
  }

  return { assistantText, tokensIn, tokensOut, cacheCreate, cacheRead };
}

// Phase labels: index 0 = setup, index 1 = sprint, extras get phase<n>
const phaseLabels = ['setup', 'sprint'];

let allAssistantText = '';
const pmPhases = [];

for (let i = 0; i < rawFiles.length; i++) {
  const label = phaseLabels[i] ?? `phase${i + 1}`;
  const result = processRawFile(rawFiles[i], pmProvider);
  allAssistantText += result.assistantText;
  pmPhases.push({
    role: `pm-${label}`,
    tokens_in: result.tokensIn,
    tokens_out: result.tokensOut,
    cache_creation_input_tokens: result.cacheCreate,
    cache_read_input_tokens: result.cacheRead,
  });
}

// Sum member telemetry from ground-truth JSONL logs
const telemetry = [...pmPhases];

function sumMemberLogs(role) {
  let in_t = 0;
  let out_t = 0;
  const dir = join(runDir, 'logs', role);
  if (existsSync(dir)) {
    for (const file of readdirSync(dir)) {
      if (file.endsWith('.jsonl')) {
        const logContent = readFileSync(join(dir, file), 'utf8');
        for (const line of logContent.split('\n')) {
          if (!line.trim()) continue;
          try {
            const entry = JSON.parse(line);
            if (entry.tokens) {
              in_t += (entry.tokens.input ?? entry.tokens.input_tokens ?? 0);
              out_t += (entry.tokens.output ?? entry.tokens.output_tokens ?? 0);
            }
          } catch {}
        }
      }
    }
  }
  return { tokens_in: in_t, tokens_out: out_t };
}

telemetry.push({ role: 'doer', ...sumMemberLogs('doer') });
telemetry.push({ role: 'reviewer', ...sumMemberLogs('reviewer') });

// Extract checkpoints: one JSON object per "CHECKPOINT:" line
let checkpoints = [];
const regex = /CHECKPOINT:\s*(\{[\s\S]*?\})/g;
let match;
while ((match = regex.exec(allAssistantText)) !== null) {
  try {
    const cp = JSON.parse(match[1]);
    if (cp && cp.id) {
      const existing = checkpoints.findIndex(c => c.id === cp.id);
      if (existing >= 0) checkpoints[existing] = cp;
      else checkpoints.push(cp);
    }
  } catch {}
}

// A phase passes only if its terminal checkpoint was emitted.
const TERMINALS = { setup: 'T2-done', sprint: 'T3-done' };
const requiredTerminals = [];
for (let i = 0; i < rawFiles.length; i++) {
  const label = phaseLabels[i];
  if (label && TERMINALS[label] && existsSync(rawFiles[i])) {
    requiredTerminals.push(TERMINALS[label]);
  }
}
const ids = new Set(checkpoints.map(c => c.id));
const missingTerminals = requiredTerminals.filter(t => !ids.has(t));
const hasFail = checkpoints.some(c => c.status === 'FAIL');
const overall = (checkpoints.length === 0 || hasFail || missingTerminals.length > 0) ? 'FAIL' : 'PASS';

const report = {
  run: {
    suite,
    pm_os:       pmOs,
    pm_provider: pmProvider,
    timestamp:   new Date().toISOString().replace(/\.\d+Z$/, 'Z'),
  },
  results: checkpoints,
  overall,
  missing_terminals: missingTerminals,
  telemetry,
};

process.stdout.write(JSON.stringify(report, null, 2) + '\n');
